src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c

```diff
@@ -4,41 +4,24 @@
 
 #include <stdbool.h>
 #include <stdint.h>
+#include <stdlib.h>
 
 #include "pal_errors_internal.h"
 #include "pal_locale_internal.h"
 #include "pal_timeZoneInfo.h"
 
-/*
-Gets the localized display name for the specified time zone.
-*/
-ResultCode GlobalizationNative_GetTimeZoneDisplayName(const UChar* localeName,
-                                                      const UChar* timeZoneId,
-                                                      TimeZoneDisplayNameType type,
-                                                      UChar* result,
-                                                      int32_t resultLength)
-{
-    UErrorCode err = U_ZERO_ERROR;
-    char locale[ULOC_FULLNAME_CAPACITY];
-    GetLocale(localeName, locale, ULOC_FULLNAME_CAPACITY, false, &err);
-
-    int32_t timeZoneIdLength = -1; // timeZoneId is NULL-terminated
-    UCalendar* calendar = ucal_open(timeZoneId, timeZoneIdLength, locale, UCAL_DEFAULT, &err);
+#define DISPLAY_NAME_LENGTH 256  // arbitrarily large, to be safe
+#define TZID_LENGTH 64           // arbitrarily large, to be safe
 
-    // TODO (https://github.com/dotnet/runtime/issues/16232): need to support Generic names, but ICU "C" api
-    // has no public option for this. For now, just use the ICU standard name for both Standard and Generic
-    // (which is the same behavior on Windows with the mincore TIME_ZONE_INFORMATION APIs).
-    ucal_getTimeZoneDisplayName(
-        calendar,
-        type == TimeZoneDisplayName_DaylightSavings ? UCAL_DST : UCAL_STANDARD,
-        locale,
-        result,
-        resultLength,
-        &err);
+// For descriptions of the following patterns, see https://unicode-org.github.io/icu/userguide/format_parse/datetime/#date-field-symbol-table
+static const UChar GENERIC_PATTERN_UCHAR[] = {'v', 'v', 'v', 'v', '\0'};           // u"vvvv"
+static const UChar GENERIC_LOCATION_PATTERN_UCHAR[] = {'V', 'V', 'V', 'V', '\0'};  // u"VVVV"
+static const UChar EXEMPLAR_CITY_PATTERN_UCHAR[] = {'V', 'V', 'V', '\0'};          // u"VVV"
 
-    ucal_close(calendar);
-    return GetResultCode(err);
-}
+// Declare some private functions, implemented later in this file.
+void GetTimeZoneDisplayName_FromCalendar(const char* locale, const UChar* timeZoneId, const UDate timestamp, UCalendarDisplayNameType type, UChar* result, int32_t resultLength, UErrorCode* err);
```

Comment:
>GetTimeZoneDisplayName_FromCalendar [](start = 5, length = 35)

why we didn't just moved the methods up and no need to declare them here?

Reply:
It just seemed cleaner to have the large private methods below.  If there's any overhead, or you just prefer them the other way, let me know and I'll move them.

Reply:
I'll just move the public method last and remove these declarations.  No big deal.